wallet: Consolidate CInputCoin and COutput

achow101 opened the pull request:
While working on coin selection code, it occurred to me that `CInputCoin` is really a subset of `COutput` and the conversion of a `COutput` to a `CInputCoin` does not appear to be all that useful. So this PR adds fields that are present in `CInputCoin` to `COutput` and replaces the usage of `CInputCoin` with `COutput`.

`COutput` is also moved to coinselection.h. As part of this move, the usage of `CWalletTx` is removed from `COutput`. It is instead replaced by storing a `COutPoint` and the `CTxOut` rather than the entire `CWalletTx` as coin selection does not really need the full `CWalletTx`. The `CWalletTx` was only used for figuring out whether the transaction containing the output was from the current wallet, and for the transaction's time. These are now parameters to `COutput`'s constructor.

achow101 committed:
wallet: Remove CWallet and CWalletTx from COutput's constructor

achow101 committed:
moveonly: move COutput to coinselection.h

achow101 committed:
coinselection: Add effective value and fees to COutput

achow101 committed:
coinselection: Use COutput instead of CInputCoin

Also rename setPresetCoins to preset_coins

achow101 committed:
coinselection: Remove CInputCoin

It is no longer needed as everything it was doing is now done by COutput

Sjors commented on src/wallet/coinselection.h:
```diff
@@ -109,6 +109,10 @@ class COutput
 
     bool m_from_me;
 
+    CAmount effective_value;
```
864b4edf130afe2b77142ca343022b3de6d3fec5: did you mean `m_ effective_value`?

achow101 replied:
Done

Sjors commented on src/wallet/coinselection.h:
```diff
@@ -109,6 +109,10 @@ class COutput
 
+    CAmount effective_value;
+    CAmount m_fee{0};
+    CAmount m_long_term_fee{0};
```
864b4edf130afe2b77142ca343022b3de6d3fec5: this is not set and not used

achow101 replied:
It is used later, but I didn't want that commit to be too big.

Sjors replied:
06cf38fce16f327579b008ee8fb78fa851cc8fce: it compiles fine now. But ideally you would move some more code from d840d84f547e911ec88f098d1ece7ed86f100cbc into this commit to set and use them. Not sure if that's practical.

Sjors commented on src/wallet/coinselection.h:
```diff
@@ -87,7 +87,7 @@ class COutput
 
     /** Pre-computed estimated size of this output as a fully-signed input in a transaction. Can be -1 if it could not be calculated */
-    int nInputBytes;
+    int m_input_bytes;
```
6d54822cbcd466df7f819430e3be87e480e41e77: did you mean to do this in the previous commit (that would make this commit a bit smaller too)?

achow101 replied:
Done

Sjors reviewed (commented):
Concept ACK

Reviewed the first 3 commits up to 95338d3b4bf3be2492cc654029ebfb28bedda42d, got a bit confused after that, as I think some of 864b4ed accidentally landed in 6d54822.

While you're touching this, it would be helpful to document the `COutput` class (now in `coinselection.h`) and how it relates to `CWalletTx` and `COutputEntry` (which is in `receive.h`, but used in the `listtransactions` RPC). For example, I was somewhat confused to see that `CWalletTx` does not in fact contain a list of `COutput`s.

95338d3b4bf3be2492cc654029ebfb28bedda42d: do you want to move `AvailableCoins`, `ListCoins`, etc from `spend.h` to `coinselection.h` too? Those all use `COutput`. Or is that more involved?

achow101 committed:
wallet: Store whether a COutput is from the wallet

Instead of determining whether the containing transaction is from the
wallet dynamically as needed, just pass it in to COutput and store it.
The transaction ownership isn't going to change.

achow101 committed:
wallet: Replace CWalletTx in COutput with COutPoint and CTxOut

Instead of having a pointer to the CWalletTx in COutput, we can just
store the COutPoint and the CTxOut as those are the only things we need
from the CWalletTx. Other things CWalletTx used to provide were time and
fIsFromMe but these are also being stored by COutput.

achow101 committed:
wallet: cleanup COutput constructor

achow101 committed:
wallet: Store tx time in COutput

achow101 committed:
wallet: Provide input bytes to COutput

achow101 commented on src/wallet/spend.h:
```diff
@@ -48,21 +47,34 @@ class COutput
+        nDepth(nDepthIn),
+        fSpendable(fSpendableIn),
+        fSolvable(fSolvableIn),
+        fSafe(fSafeIn),
```
Fixed

achow101 commented on src/wallet/spend.h:
```diff
@@ -48,21 +47,34 @@ class COutput
-    COutput(const CWallet& wallet, const CWalletTx& wtx, int iIn, int nDepthIn, bool fSpendableIn, bool fSolvableIn, bool fSafeIn, bool use_max_sig_in = false)
+    int64_t m_time;
+
+    bool m_from_me;
```
Done

achow101 commented:
I've split up a few commits and introduced the variables added to `COutput` in more separate commits.

> do you want to move `AvailableCoins`, `ListCoins`, etc from `spend.h` to `coinselection.h` too? Those all use `COutput`. Or is that more involved?

No. `COutput` is only being moved to avoid a circular dependency.

maflcko reviewed (commented):
Concept ACK. I need this for some wallet stuff.

kiminuo commented:
Just read commits up to 1f475858590a50a8a3f37a02df2c2d7b803f9834 (*wallet: Remove CWalletTx from COutput's constructor*) and so far I have not found any obvious issue. Nice PR.

promag reviewed (commented):
Light code review ACK 2278038d3f9c2488421cf7ed1131ac77a6f22380.

nit, drop `COutput` forward declaration in wallet.h

instagibbs commented:
concept ACK

Sjors commented on src/wallet/spend.h:
```diff
@@ -52,9 +52,9 @@ class COutput
     /** Whether the transaction containing this output is sent from the owning wallet */
     bool m_from_me;
 
-    COutput(const CWallet& wallet, const CWalletTx& wtx, int iIn, int depth, int input_bytes, bool spendable, bool solvable, bool safe, int64_t time, bool from_me)
```
1f475858590a50a8a3f37a02df2c2d7b803f9834: it seems `wallet` was already unused at this point, so you could drop it in another commit. But at least the commit message should mention them both.

Sjors commented on src/bench/coin_selection.cpp:
```diff
@@ -82,9 +82,9 @@ static void add_coin(const CAmount& nValue, int nInput, std::vector<OutputGroup>
     tx.vout.resize(nInput + 1);
     tx.vout[nInput].nValue = nValue;
-    CInputCoin coin(MakeTransactionRef(tx), nInput);
+    COutput output(COutPoint(tx.GetHash(), nInput), tx.vout.at(nInput), 0, -1, true, true, true, 0, true);
```
d840d84f547e911ec88f098d1ece7ed86f100cbc: document booleans?

Sjors reviewed (commented):
Reviewed everything except d840d84f547e911ec88f098d1ece7ed86f100cbc, which still gives me a bit of a headache, but will look at it later.

Sjors commented on src/wallet/spend.cpp:
```diff
@@ -432,10 +430,10 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, const std::vec
              * positive_only is set to false because we want to include all preset inputs, even if they are dust.
              */
-            preset_inputs.Insert(out.GetInputCoin(), 0, false, 0, 0, false);
+            preset_inputs.Insert(out, 0, 0, false);
```
nit: document the remaining bools while touching this

achow101 replied:
Done

Sjors reviewed (approved):
utACK 3f5ed1858eba99253153f6b692625af075107f60

instagibbs reviewed (approved):
ACK 3f5ed1858eba99253153f6b692625af075107f60

Xekyo commented:
Concept ACK

kallewoof commented on src/wallet/coinselection.cpp:
```diff
@@ -50,8 +50,8 @@ struct {
- * @param const std::vector<CInputCoin>& utxo_pool The set of UTXOs that we are choosing from.
- *        These UTXOs will be sorted in descending order by effective value and the CInputCoins'
+ * @param const std::vector<OutputGroup>& utxo_pool The set of UTXOs that we are choosing from.
+ *        These UTXOs will be sorted in descending order by effective value and the OutputGroups'
```
In 18a6e09f0a50736a8113b6e6d918d93aa19d5a22

Maybe `The set of UTXO groups that we are choosing from` / `These UTXO groups will be sorted in[...]` is more accurate now.

achow101 replied:
Done

kallewoof commented on src/wallet/spend.cpp:
```diff
@@ -470,27 +468,28 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, const std::vec
     for (std::vector<COutput>::iterator it = vCoins.begin(); it != vCoins.end() && coin_control.HasSelected();)
     {
-        if (setPresetCoins.count(it->GetInputCoin()))
+        if (setPresetCoins.count(it->m_outpoint))
```
In 18a6e09f0a50736a8113b6e6d918d93aa19d5a22

Style/nit: It is perfectly possible to do `s/setPresetCoins/preset_coins` in this commit without changing diff count, as it touches all references to `setPresetCoins`.

achow101 replied:
Done

kallewoof reviewed (commented):
Concept/utACK 3f5ed1858eba99253153f6b692625af075107f60

taveechaimekwan commented:
Khlong Nueng Subdistrict 12130

ryanofsky commented on src/wallet/spend.h:
```diff
@@ -51,7 +51,10 @@ class COutput
+    /** Whether the transaction containing this output is sent from the owning wallet */
+    bool m_from_me;
+
+    COutput(const CWallet& wallet, const CWalletTx& wtx, int iIn, int depth, bool spendable, bool solvable, bool safe, int64_t time, bool from_me, bool use_max_sig_in = false)
```
In commit "wallet: Store whether a COutput is from the wallet" (521a703d056c13d85dbaabd289d2cc7edfd671c0)

Again `use_max_sig_in = false` default argument makes this hard to verify and means it could silently break. This again suggests dropping default argument in earlier commit.

ryanofsky commented on src/wallet/spend.cpp:
```diff
@@ -162,6 +162,8 @@ void AvailableCoins(const CWallet& wallet, std::vector<COutput>& vCoins, const C
             continue;
         }
 
+        bool tx_from_me = CachedTxIsFromMe(wallet, wtx, ISMINE_ALL);
```
In commit "wallet: Store whether a COutput is from the wallet" (521a703d056c13d85dbaabd289d2cc7edfd671c0)

This seems ok, but just reading commit description and looking at this code this doesn't seem to be an improvement. Two CachedTxIsFromMe calls are added and two are dropped which is neutral, but COutput parameter list is growing even more, which seems bad.

I'm assuming something justifies this in a later commit, but it would be good to hint at whatever this might be in commit description,

EDIT: After reviewing PR, still not exactly clear what benefit of this change is. But seems basically neutral.

achow101 replied:
The purpose is that when filtering UTXOs, we need to know whether they belong to the wallet. Since we will be removing access to CWallet and CWalletTx in the future, this information must be stored in COutput itself.

ryanofsky commented on src/wallet/spend.h:
```diff
@@ -54,24 +51,17 @@ class COutput
     bool m_from_me;
 
-    COutput(const CWallet& wallet, const CWalletTx& wtx, int iIn, int depth, bool spendable, bool solvable, bool safe, int64_t time, bool from_me, bool use_max_sig_in = false)
+    COutput(const CWallet& wallet, const CWalletTx& wtx, int iIn, int depth, int input_bytes, bool spendable, bool solvable, bool safe, int64_t time, bool from_me)
```
In commit "wallet: Provide input bytes to COutput" (353d7bf8e7c800aa19cbcc52fe0636667c754cec)

Nice to drop the use_max_sig argument here, but this should definitely drop the CWallet argument too, which is no longer needed here

EDIT: I see it is dropped in later commit

ryanofsky reviewed (approved):
Code review ACK 3f5ed1858eba99253153f6b692625af075107f60. All my suggestions are minor, and you can ignore them. This is a nice simplification and the changes are generally pretty easy to follow. Main feedback would be to drop the struct `m_` prefixes (Ugly!)

achow101 commented on src/wallet/spend.h:
```diff
@@ -48,15 +48,19 @@ class COutput
     bool m_safe;
 
-    COutput(const CWallet& wallet, const CWalletTx& wtx, int iIn, int depth, bool spendable, bool solvable, bool safe, bool use_max_sig_in = false)
+    /** The time of the transaction containing this output */
```
It's whatever the CWalletTx reports as its time. It's not entirely clear what that is, and depends on `nTimeSmart` being... smart.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -195,7 +195,7 @@ void AvailableCoins(const CWallet& wallet, std::vector<COutput>& vCoins, const C
             bool spendable = ((mine & ISMINE_SPENDABLE) != ISMINE_NO) || (((mine & ISMINE_WATCH_ONLY) != ISMINE_NO) && (coinControl && coinControl->fAllowWatchOnly && solvable));
 
-            vCoins.push_back(COutput(wallet, wtx, i, nDepth, spendable, solvable, safeTx, (coinControl && coinControl->fAllowWatchOnly)));
+            vCoins.emplace_back(wallet, wtx, i, nDepth, spendable, solvable, safeTx, wtx.GetTxTime(), (coinControl && coinControl->fAllowWatchOnly));
```
That gets dropped in a later commit anyways.

achow101 replied:
Done anyways.

achow101 committed:
scripted-diff: Rename COutput member variables

Update the member variables to match the new style

-BEGIN VERIFY SCRIPT-
sed -i 's/fSpendableIn/spendable/' $(git grep -l "fSpendableIn")
sed -i 's/fSpendable/spendable/' $(git grep -l "fSpendable")
sed -i 's/fSolvableIn/solvable/' $(git grep -l "fSolvableIn")
sed -i 's/fSolvable/solvable/' $(git grep -l "fSolvable")
sed -i 's/fSafeIn/safe/' $(git grep -l "fSafeIn")
sed -i 's/fSafe/safe/' $(git grep -l "fSafe")
sed -i 's/nInputBytes/input_bytes/' $(git grep -l "nInputBytes")
sed -i 's/nDepthIn/depth/' $(git grep -l "nDepthIn" src/wallet src/bench)
sed -i 's/nDepth/depth/' src/wallet/spend.h
sed -i 's/\.nDepth/.depth/' $(git grep -l "\.nDepth" src/wallet/)
sed -i 's/nDepth, FormatMoney/depth, FormatMoney/' src/wallet/spend.cpp
-END VERIFY SCRIPT-

achow101 committed:
wallet: Remove use_max_sig default value

As we change the constructor for COutput, it becomes somewhat dangerous
if there are default values.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -470,27 +468,28 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, const std::vec
-        if (coin.m_input_bytes == -1) {
+        /* Set some defaults for depth, spendable, solvable, safe, time, and from_me as these don't matter for preset inputs since no selection is being done. */
+        COutput output(outpoint, txout, /*depth=*/ 0, input_bytes, /*spendable=*/ true, /*solvable=*/ true, /*safe=*/ true, /*time=*/ 0, /*from_me=*/ false);
+        if (output.m_input_bytes == -1) {
```
Done

achow101 replied:
Actually we still need this check here because `input_bytes` could be set by either `CalculateMaximumSignedInputSize` or `GetVirtualTransactionSize`, depending on whether the input weight is provided (this may have been a hidden merge conflict though so not apparent in the diff).

achow101 commented on src/wallet/spend.h:
```diff
@@ -27,16 +27,16 @@ class COutput
      * If = 0: the tx is waiting confirmation.
      * If < 0: a conflicting tx is on chain and has this many confirmations. */
-    int nDepth;
+    int m_depth;
```
Dropped the `m_` prefix.

achow101 commented on src/wallet/spend.h:
```diff
@@ -48,15 +48,19 @@ class COutput
+    /** The time of the transaction containing this output */
+    int64_t m_time;
+
+    COutput(const CWallet& wallet, const CWalletTx& wtx, int iIn, int depth, bool spendable, bool solvable, bool safe, int64_t time, bool use_max_sig_in = false)
```
I've added an intermediate commit removing the default value.

Sjors commented:
CI seems a bit unhappy

ryanofsky commented on src/wallet/coinselection.cpp:
```diff
@@ -50,8 +50,8 @@ struct {
  *
- * @param const std::vector<CInputCoin>& utxo_pool The set of UTXOs that we are choosing from.
- *        These UTXOs will be sorted in descending order by effective value and the CInputCoins'
+ * @param const std::vector<OutputGroup>& utxo_pool The set of UTXO groups that we are choosing from.
+ *        These UTXO groups will be sorted in descending order by effective value and the OutputGroups'
```
In commit "coinselection: Use COutput instead of CInputCoin" (f2cf89eefefec98aa6f5f7a8d3f6f0abc0e0d526)

Note: parameter isn't actually changing here, obsolete comment is just being brought up to date with current code. (At first I thought this change was more significant than it really is)

ryanofsky commented on src/wallet/coinselection.h:
```diff
@@ -133,6 +133,18 @@ class COutput
+
+    bool operator==(const COutput& rhs) const {
+        return outpoint == rhs.outpoint;
+    }
```
In commit "coinselection: Use COutput instead of CInputCoin" (f2cf89eefefec98aa6f5f7a8d3f6f0abc0e0d526)

Minor: operator< above makes sense I think, but these operator== operator!= methods are only used by tests and don't really describe full equality. I'd suggest dropping these == != methods and just passing a lambda to std::mismatch in the test so it can explicitly specify the equality it checks for.

achow101 replied:
Done in an additional commit.

ryanofsky reviewed (approved):
Code review ACK 1a3e984a09343ac1d804f6106a52c9cf527f7dba

Nice simplification and easy review!

Sjors reviewed (approved):
re-utACK 1a3e984a09343ac1d804f6106a52c9cf527f7dba

glozow reviewed (commented):
light code review ACK 1a3e984a09343ac1d804f6106a52c9cf527f7dba, just a few nits

glozow commented on src/wallet/spend.h:
```diff
@@ -46,30 +46,30 @@ class COutput
-        nInputBytes(-1),
-        fSpendable(fSpendableIn),
-        fSolvable(fSolvableIn),
+        depth(depth),
```
Ah I missed that, thanks :+1:

Xekyo commented on src/wallet/rpc/coins.cpp:
```diff
@@ -648,16 +648,16 @@ RPCHelpMan listunspent()
-        entry.pushKV("txid", out.tx->GetHash().GetHex());
-        entry.pushKV("vout", out.i);
+        entry.pushKV("txid", out.outpoint.hash.GetHex());
+        entry.pushKV("vout", (int)out.outpoint.n);
```
Isn't `out.i` gone after this commit?

achow101 replied:
`out.i` is removed by this PR.

Xekyo commented on src/wallet/spend.cpp:
```diff
@@ -467,27 +465,29 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, const std::vec
+
+        /* Set some defaults for depth, spendable, solvable, safe, time, and from_me as these don't matter for preset inputs since no selection is being done. */
+        COutput output(outpoint, txout, /*depth=*/ 0, input_bytes, /*spendable=*/ true, /*solvable=*/ true, /*safe=*/ true, /*time=*/ 0, /*from_me=*/ false);
+        output.effective_value = output.txout.nValue - coin_selection_params.m_effective_feerate.GetFee(output.input_bytes);
```
Probably for a follow-up, but I think 

`COutput` could have a function that takes a feerate, and sets the `COutput.effective_value`, or perhaps we could even have `COutput` take a feerate in the constructor.

achow101 replied:
I'll leave that for a followup.

murchandamus reviewed (commented):
ACK, thanks for doing, I'm glad to see CInputCoin go away. Got some nits and questions.

w0xlt reviewed (commented):
Code Review ACK https://github.com/bitcoin/bitcoin/pull/24091/commits/1a3e984a09343ac1d804f6106a52c9cf527f7dba

achow101 committed:
coinselection: Remove COutput operators == and !=

These operators are used only by the tests in std::mismatch. As
std::mismatch can take a binary predicate, we can use a lambda that
achieves the same instead.

achow101 commented on src/wallet/spend.h:
```diff
@@ -49,8 +49,15 @@ class COutput
+        fSpendable(fSpendableIn),
+        fSolvable(fSolvableIn),
+        use_max_sig(use_max_sig_in),
+        fSafe(fSafeIn)
```
I would, but it causes conflicts with every other commit in this PR. The effort is not worth it.

achow101 commented on src/wallet/interfaces.cpp:
```diff
@@ -111,6 +111,17 @@ WalletTxOut MakeWalletTxOut(const CWallet& wallet,
     return result;
 }
 
+WalletTxOut MakeWalletTxOut(const CWallet& wallet,
```
`MakeWalletTxOut` is used to create entries for the GUI. Part of that is the transaction's time, retrieved with `CWalletTx::GetTxTime`. This constructor is called for each `COutput` returned by `ListCoins`. As `COutput` will no longer have access to the `CWalletTx` later in this PR, it is necessary for it to be able to have access to the time so that `WalletTxOut` can have the correct time set. That is why this change exists, and it is, in fact, related to storing time in the `COutput`.

achow101 commented on src/wallet/spend.h:
```diff
@@ -38,9 +38,6 @@ class COutput
     /** Whether we know how to spend this output, ignoring the lack of keys */
     bool solvable;
 
-    /** Whether to use the maximum sized, 72 byte signature when calculating the size of the input spend. This should only be set when watch-only outputs are allowed */
```
Moved the comment.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -191,7 +193,7 @@ void AvailableCoins(const CWallet& wallet, std::vector<COutput>& vCoins, const C
             bool spendable = ((mine & ISMINE_SPENDABLE) != ISMINE_NO) || (((mine & ISMINE_WATCH_ONLY) != ISMINE_NO) && (coinControl && coinControl->fAllowWatchOnly && solvable));
 
-            vCoins.emplace_back(wallet, wtx, i, nDepth, spendable, solvable, safeTx, wtx.GetTxTime(), /*use_max_sig_in=*/ (coinControl && coinControl->fAllowWatchOnly));
+            vCoins.emplace_back(wallet, wtx, i, nDepth, spendable, solvable, safeTx, wtx.GetTxTime(), tx_from_me, /*use_max_sig_in=*/ (coinControl && coinControl->fAllowWatchOnly));
```
I would, but it causes conflicts with most of the other commit in this PR. The effort is not worth it.

achow101 commented on src/bench/coin_selection.cpp:
```diff
@@ -58,7 +58,7 @@ static void CoinSelection(benchmark::Bench& bench)
     std::vector<COutput> coins;
     for (const auto& wtx : wtxs) {
-        coins.emplace_back(wallet, *wtx, /*iIn=*/ 0, /*depth=*/ 6 * 24, /*spendable=*/ true, /*solvable=*/ true, /*safe=*/ true, wtx->GetTxTime(), /*from_me=*/ true, /*use_max_sig_in=*/ false);
+        coins.emplace_back(wallet, *wtx, /*iIn=*/ 0, /*depth=*/ 6 * 24, GetTxSpendSize(wallet, *wtx, 0), /*spendable=*/ true, /*solvable=*/ true, /*safe=*/ true, wtx->GetTxTime(), /*from_me=*/ true);
```
I would, but it causes conflicts with every other commit in this PR. The effort is not worth it.

achow101 commented on src/wallet/coinselection.h:
```diff
@@ -110,6 +110,10 @@ class COutput
     /** Whether the transaction containing this output is sent from the owning wallet */
     bool from_me;
 
+    CAmount effective_value;
+    CAmount fee{0};
+    CAmount long_term_fee{0};
```
Done

ryanofsky reviewed (approved):
Code review ACK 049003fe68a4183f6f20da16f58f10079d1e02df, just adding comments and removing == operators since last review

w0xlt reviewed (approved):
reACK 049003f

Xekyo commented:
reACK 049003fe68a4183f6f20da16f58f10079d1e02df

murchandamus reviewed (approved):
reACK 049003fe68a4183f6f20da16f58f10079d1e02df
